File: library/src/main/java/me/shkschneider/skeleton/helper/NotificationHelper.java
package me.shkschneider.skeleton.helper;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.v4.app.NotificationCompat;

/**
 * +------+------------------------------+
 * |      | contentTitle       timestamp |
 * | ICON | contentText                  |
 * |      | subText          contentInfo |
 * +------+------------------------------+
 * API-21+
 * +----------+--------------------------+
 * |          | contentTitle   timestamp |
 * |   ICON   | contentText              |
 * |     icon | subText      contentInfo |
 * +----------+--------------------------+
 * API-24+
 * +-------------------------------------+
 * | icon app - subText - timestamp      |
 * | contentTitle                        |
 * | contentText                         |
 * +-------------------------------------+
 */
public class NotificationHelper {

    public static PendingIntent pendingIntent(@NonNull final Activity activity, Intent intent) {
        if (intent == null) {
            intent = new Intent(activity, activity.getClass());
        }
        if (intent.getFlags() == 0) {
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        }
        return PendingIntent.getActivity(activity, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    public static class Builder extends NotificationCompat.Builder {

        public Builder(@NonNull final Channel channel) {
            super(ContextHelper.applicationContext(), channel.id);
            if (AndroidHelper.api() >= AndroidHelper.API_26) {
                notificationChannel26(channel);
            }
        }

        @Deprecated
        public Builder(@NonNull final String id) {
            super(ContextHelper.applicationContext(), id);
            // setShowWhen(AndroidHelper.api() < AndroidHelper.API_24);
            // setVisibility(NotificationCompat.VISIBILITY_PUBLIC);
            // setPriority(NotificationCompat.PRIORITY_DEFAULT);
            // setDefaults(NotificationCompat.DEFAULT_ALL);
            // setOngoing(false);
            // setAutoCancel(true);
        }

        @TargetApi(AndroidHelper.API_26)
        private void notificationChannel26(@NonNull final Channel channel) {
            if (Channel.get(channel.id) != null) {
                return;
            }
            final NotificationChannel notificationChannel = new NotificationChannel(channel.id, channel.name, NotificationManager.IMPORTANCE_DEFAULT);
            notificationChannel.setLockscreenVisibility(NotificationCompat.VISIBILITY_PUBLIC);
            notificationChannel.setShowBadge(channel.badge);
            notificationChannel.enableLights(channel.lights);
            notificationChannel.enableVibration(channel.vibration);
            SystemServices.notificationManager().createNotificationChannel(notificationChannel);
        }

        public Notification build() {
            return super.build();
        }

    }

    public static class Channel {

        public String id;
        public String name;
        public boolean badge;
        public boolean lights;
        public boolean vibration;

        public Channel(@NonNull final String id, @NonNull final String name) {
            this(id, name, true, true, true);
        }

        public Channel(@NonNull final String id, @NonNull final String name,
                       final boolean badge, final boolean lights, final boolean vibration) {
            this.id = id;
            this.name = name;
            this.badge = badge;
            this.lights = lights;
            this.vibration = vibration;
        }

        @TargetApi(AndroidHelper.API_26)
        public NotificationChannel get() {
            NotificationChannel notificationChannel = get(this.id);
            if (notificationChannel != null) {
                return notificationChannel;
            }
            notificationChannel = new NotificationChannel(this.id, this.name, NotificationManager.IMPORTANCE_DEFAULT);
            notificationChannel.setLockscreenVisibility(NotificationCompat.VISIBILITY_PUBLIC);
            notificationChannel.setShowBadge(this.badge);
            notificationChannel.enableLights(this.lights);
            notificationChannel.enableVibration(this.vibration);
            return notificationChannel;
        }

        @TargetApi(AndroidHelper.API_26)
        public static NotificationChannel get(@NonNull final String id) {
            return SystemServices.notificationManager().getNotificationChannel(id);
        }

        @TargetApi(AndroidHelper.API_26)
        public static void create(@NonNull final NotificationChannel notificationChannel) {
            SystemServices.notificationManager().createNotificationChannel(notificationChannel);
        }

        @TargetApi(AndroidHelper.API_26)
        public static void delete(@NonNull final String id) {
            SystemServices.notificationManager().deleteNotificationChannel(id);
        }

    }

    public static void notify(@IntRange(from=0) final int id, @NonNull final Notification notification) {
        // final String channelId = NotificationCompat.getChannelId(notification);
        notification.when = System.currentTimeMillis(); // ?
        SystemServices.notificationManager().notify(id, notification);
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/helper/NotificationHelper.kt
package me.shkschneider.skeleton.helper

import android.app.*
import android.content.Intent
import android.os.Build
import android.support.annotation.IntRange
import android.support.annotation.RequiresApi
import android.support.v4.app.NotificationCompat
import java.util.concurrent.TimeUnit

/**
 * +------+------------------------------+
 * |      | contentTitle       timestamp |
 * | ICON | contentText                  |
 * |      | subText          contentInfo |
 * +------+------------------------------+
 * API-21+
 * +----------+--------------------------+
 * |          | contentTitle   timestamp |
 * |   ICON   | contentText              |
 * |     icon | subText      contentInfo |
 * +----------+--------------------------+
 * API-24+
 * +-------------------------------------+
 * | icon app - subText - timestamp      |
 * | contentTitle                        |
 * | contentText                         |
 * +-------------------------------------+
 */
object NotificationHelper {

    fun pendingIntent(activity: Activity, intent: Intent): PendingIntent {
        if (intent.flags == 0) {
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        return PendingIntent.getActivity(activity, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)
    }

    fun pendingIntent(activity: Activity): PendingIntent {
        return pendingIntent(activity, Intent(activity, activity::class.java))
    }

    class Builder : NotificationCompat.Builder {

        constructor(channel: Channel) : super(ContextHelper.applicationContext(), channel.id) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                Channel.get(channel.id) ?: return
                with(NotificationChannel(channel.id, channel.name, NotificationManager.IMPORTANCE_DEFAULT)) {
                    lockscreenVisibility = NotificationCompat.VISIBILITY_PUBLIC
                    setShowBadge(channel.badge)
                    enableLights(channel.lights)
                    enableVibration(channel.vibration)
                    SystemServices.notificationManager()?.createNotificationChannel(this)
                }
            }
            // For apps targeting N and above, this time is not shown anymore by default.
            if (AndroidHelper.api() < AndroidHelper.API_24) {
                setShowWhen(false)
            }
            // OnClick
            setAutoCancel(true)
        }

        override fun setOngoing(ongoing: Boolean): NotificationCompat.Builder {
            if (ongoing) {
                setShowWhen(false)
            }
            return super.setOngoing(ongoing)
        }

        fun setWhenMilliseconds(ms: Long): NotificationCompat.Builder {
            return super.setWhen(ms)
        }

        fun setWhenSeconds(s: Long): NotificationCompat.Builder {
            return super.setWhen(TimeUnit.SECONDS.toMillis(s))
        }

        @Deprecated("Use a NotificationChannel.")
        constructor(id: String) : super(ContextHelper.applicationContext(), id) {
            setChannelId(id)
            // setShowWhen(AndroidHelper.api() < AndroidHelper.API_24);
            // setVisibility(NotificationCompat.VISIBILITY_PUBLIC);
            // setPriority(NotificationCompat.PRIORITY_DEFAULT);
            // setDefaults(NotificationCompat.DEFAULT_ALL);
            // setOngoing(false);
            // setAutoCancel(true);
        }

        override fun build(): Notification {
            return super.build()
        }

    }

    class Channel {

        val id: String
        val name: String
        val badge: Boolean
        val lights: Boolean
        val vibration: Boolean

        constructor(id: String, name: String, badge: Boolean = true, lights: Boolean = true, vibration: Boolean = true) {
            this.id = id
            this.name = name
            this.badge = badge
            this.lights = lights
            this.vibration = vibration
        }

        @RequiresApi(AndroidHelper.API_26)
        fun get(): NotificationChannel {
            get(id)?.let {
                return it
            }
            val notificationChannel = NotificationChannel(id, name, NotificationManager.IMPORTANCE_DEFAULT)
            with (notificationChannel) {
                lockscreenVisibility = NotificationCompat.VISIBILITY_PUBLIC
                setShowBadge(badge)
                enableLights(lights)
                enableVibration(vibration)
                return this
            }
        }

        companion object {

            @RequiresApi(AndroidHelper.API_26)
            fun get(id: String): NotificationChannel? {
                return SystemServices.notificationManager()?.getNotificationChannel(id)
            }

            @RequiresApi(AndroidHelper.API_26)
            fun create(notificationChannel: NotificationChannel) {
                SystemServices.notificationManager()?.createNotificationChannel(notificationChannel)
            }

            @RequiresApi(AndroidHelper.API_26)
            fun delete(id: String) {
                SystemServices.notificationManager()?.deleteNotificationChannel(id)
            }

        }

    }

    fun notify(@IntRange(from = 0) id: Int, notification: Notification, ms: Long? = null) {
        ms?.let {
            notification.`when` = ms
        }
        SystemServices.notificationManager()?.notify(id, notification)
    }

}
